### ISSUE TITLE
Handle N-only batches in hg38 pre-training

### ISSUE THREAD

**Issue by ravwojdyla** [kuleshov-group/caduceus#77]
It is our understanding that:
 * hg38 pre-training data contains only the standard 4 nucleotides as well as the special `N` (a non-specific nucleotide)
 * the hg38 dataset reader replaces `N` with pad token [here](https://github.com/kuleshov-group/caduceus/blob/0060a6d8079b6a040fc55d505e15972a327b70a6/src/dataloaders/datasets/hg38_dataset.py#L211-L212)
 * the pad tokens are ignored in the CE loss due to `ignore_index` config [here](https://github.com/kuleshov-group/caduceus/blob/0060a6d8079b6a040fc55d505e15972a327b70a6/configs/experiment/hg38/hg38.yaml#L11)
 * it is possible to have a pre-training batch of **only** `N` tokens in the hg38 data
   * let's call this case the "N-only batch"
   * I tested this on $$2^{17}$$ sequence length
     * if I evenly divide hg38 into 23,125 "samples" of $$2^{17}$$ sequence length, there are 816 (~3%) N-only batches

I tried to find code that would handle the N-only batch, and I could not find it. I then tried to artificially feed the training loop with a N-only batch, and the loss came out `nan`. This may lead to unintended consequences as well as confusion in reproduction, potential example https://github.com/kuleshov-group/caduceus/issues/71.

**Comment by yair-schiff** [comment]
This is an interesting find. I guess given that we used certain pre-defined splits (i.e., those from [Enformer](https://www.nature.com/articles/s41592-021-01252-x)) this did not come up in our work.

One potential workaround is to add some logic in the dataset init that drops any `N-only` rows.

**Comment by ravwojdyla** [comment]
> I guess given that we used certain pre-defined splits (i.e., those from [Enformer](https://www.nature.com/articles/s41592-021-01252-x)) this did not come up in our work.

👋 @yair-schiff is that handled by the [HG38Dataset](https://github.com/kuleshov-group/caduceus/blob/0060a6d8079b6a040fc55d505e15972a327b70a6/src/dataloaders/datasets/hg38_dataset.py#L92C7-L92C18) or something else?

**Comment by yair-schiff** [comment]
Yes, the Enformer splits are downloaded using the links for the `.bed` file in the README. Then we create a dataframe that contains the start positions of each item. The fasta file is then read to retrieve the sequences. Let me know if this is still not clear.

**Comment by ravwojdyla** [comment]
@yair-schiff ah - perfect, thank you! I can confirm that [HG38Dataset](https://github.com/kuleshov-group/caduceus/blob/0060a6d8079b6a040fc55d505e15972a327b70a6/src/dataloaders/datasets/hg38_dataset.py#L92C7-L92C18) never reads `N-only` batches, the batches may still contain `N` nucleotides, but the max number of `N`s in a single example is 2k in a $$2^{17}$$ sequence examples.

**Comment by ravwojdyla** [comment]
@yair-schiff actually, I think I spoke too soon. The original Basenji batches (which are of length $$2^{17}$$) do not suffer from the `N-only` issue (this is what I checked in the comment above). But caduceus updates/expands the Basenji intervals to length $$2^{20}$$ [here](https://github.com/kuleshov-group/caduceus/blob/0060a6d8079b6a040fc55d505e15972a327b70a6/src/dataloaders/datasets/hg38_dataset.py#L143-L144), and produces non-overlapping batches of the requested sequence length (e.g. $$2^{17}$$). So caduceus batches are not necessarily produced from the same spans as the Basenji. And it seems like there are 4 `N-only` batches (out of 272,168) in the `train` split produced by caduceus [HG38Dataset](https://github.com/kuleshov-group/caduceus/blob/0060a6d8079b6a040fc55d505e15972a327b70a6/src/dataloaders/datasets/hg38_dataset.py#L92C7-L92C18).

**Comment by yair-schiff** [comment]
We did not use 2^20 sequences for training, we only went up to 2^17. 

The dataset is structured as follows:
- Chromosome
- Start bp
- End bp (= start bp + 2^20)

When we get items from the dataset, we take non-overlapping indices within each 2^20 segment. So a full epoch would have `# of start splits from basenji` * `8` number of total sequences

**Comment by ravwojdyla** [comment]
@yair-schiff your description of the pre-processing - expansion to $$2^{20}$$ segments with smaller non-overlapping training sequences (i.e. $$8 * 2^{17}$$ per segment) - is consistent with my understanding.

This expansion and tiling process is the direct cause of the N-only samples. While the original, shorter Basenji segments may not have this characteristic, the expanded segments can cover genomic regions that result in N-only tiles/samples.

From the paper (emphasis is mine):

> We provide a more detailed description of the dataset and training methodology used in the human reference genome pre-training task. This dataset is based on the splits used in the previous Enformer study (Avsec et al., 2021). The training split comprises 34,021 segments that **we extend to a maximum length of 1,048,576 ($$2^{20}$$), collectively covering the genome** and amounting to around 35 billion tokens, or nucleotide base pairs.

I'm curious - given that the models were trained on sequences up to length $$2^{17}$$, what was the motivation for the initial expansion to $$2^{20}$$ followed by tiling (i.e. instead of training on the original $$2^{17}$$ sequences from Basenji)? Was this intended to support experiments with longer sequence lengths?

**Comment by yair-schiff** [comment]
> Was this intended to support experiments with longer sequence lengths?

Yes precisely. We initially considered going up to this maximum input length, but given computational considerations and the fact that 2^17 brought us roughly within the same input size of the popular Enformer baseline, we ended up using a max seq length of 2^17
